### PR TITLE
Use MapOutputTracker generation for failures in DAGScheduler and test failures more

### PR DESCRIPTION
**Pull request by charlesreiss**
Use MapOutputTracker generation instead of a blacklist of nodes in DAGScheduler to handle failed shuffle results from failed nodes.

Also, increment MapOutputTracker generation for all shuffle tasks, avoiding situations where an map output list with holes is cached. (This fix is suboptimal since it may result in us clearing cached map statuses more often than necessary and more often than the existing code; alternatives welcome.)

Also, add several tests to DistributedSuite which inject failures, including repeated node failures involving shuffles.

### CODE REVIEW

File: core/src/test/scala/spark/DistributedSuite.scala [hunk 1]
```diff
@@ -188,4 +188,76 @@ class DistributedSuite extends FunSuite with ShouldMatchers with BeforeAndAfter
     val values = sc.parallelize(1 to 2, 2).map(x => System.getenv("TEST_VAR")).collect()
     assert(values.toSeq === Seq("TEST_VALUE", "TEST_VALUE"))
   }
+
+  test("recover from node failures") {
+    import DistributedSuite.{markNodeIfIdentity, failOnMarkedIdentity}
+    DistributedSuite.amMaster = true
+    sc = new SparkContext(clusterUrl, "test")
+    val data = sc.parallelize(Seq(true, true), 2)
+    val singleton = sc.parallelize(Seq(true), 1)
```
Comment (mateiz) [inline core/src/test/scala/spark/DistributedSuite.scala]: Why do you create these RDDs? They seem unused.

File: core/src/main/scala/spark/scheduler/DAGScheduler.scala [hunk 2]
```diff
@@ -439,9 +445,16 @@ class DAGScheduler(taskSched: TaskScheduler) extends TaskSchedulerListener with
...
+                // recomputation of the map outputs. In that case, some nodes may have cached
+                // locations with holes (from when we detected the error) and will need the
+                // generation incremented to refetch them.
+                // TODO: Only increment the generation number if this is not the first time
+                //       we registered these map outputs.
                 mapOutputTracker.registerMapOutputs(
                   stage.shuffleDep.get.shuffleId,
-                  stage.outputLocs.map(list => if (list.isEmpty) null else list.head).toArray)
+                  stage.outputLocs.map(list => if (list.isEmpty) null else list.head).toArray,
+                  true)
```
Comment (mateiz) [inline core/src/main/scala/spark/scheduler/DAGScheduler.scala:458]: Spacing looks weird here

File: core/src/main/scala/spark/scheduler/DAGScheduler.scala [hunk 3]
```diff
@@ -72,8 +72,12 @@ class DAGScheduler(taskSched: TaskScheduler) extends TaskSchedulerListener with
   val cacheTracker = env.cacheTracker
   val mapOutputTracker = env.mapOutputTracker
 
-  val deadHosts = new HashSet[String]  // TODO: The code currently assumes these can't come back;
-                                       // that's not going to be a realistic assumption in general
+  // For tracking failed nodes, we use the MapOutputTracker's generation number, which is
+  // sent with every task. When we detect a node failing, we note the current generation number
+  // and failed host, increment it for new tasks, and use this to ignore stray ShuffleMapTask
+  // results.
+  // TODO: Garbage collect information about failure generations when new stages start.
```
Comment (mateiz) [inline core/src/main/scala/spark/scheduler/DAGScheduler.scala]: I'm not sure I understand this; couldn't a node still be down if a new stage starts, and couldn't we want to recompute some old RDD later in a different job?

File: core/src/test/scala/spark/DistributedSuite.scala (hunk 1 above)
Comment (charlesreiss) [inline core/src/test/scala/spark/DistributedSuite.scala]: Removed (and below).

File: core/src/main/scala/spark/scheduler/DAGScheduler.scala (hunk 3 above)
Comment (charlesreiss) [inline core/src/main/scala/spark/scheduler/DAGScheduler.scala]: Changed TODO to hopefully make it clearer.

**Comment by mateiz** [comment]
I've merged this manually. It conflicted slightly with your previous pull request.

**Pull request merged**